FIX #1914 Reject command lines with trailing flags

dgageot opened the pull request.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "1914-create-ignores-trailing-options" git@github.com:dgageot/machine.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

dmp42 commented:
@dgageot some integration tests are apparently relying on trailing flags - you will have to update them (see `test/integration`)

dmp42 commented:
Also, github will magically link to an issue if fix #1914 is in the body of a comment - unfortunately, not in the title of the PR.

dgageot commented:
@nathanleclaire @dmp42 wdyt?

dmp42 commented on test/integration/cli/create-rm.bats:
```diff
@@ -53,9 +53,9 @@ load ${BASE_TEST_DIR}/helpers.bash
 }
 
 @test "none: extraneous argument is ignored in name 'machine create -d none --url none a foo'" {
```
Shall this test be renamed?

dmp42 commented:
Looks good.
@nathanleclaire can you take a quick look?

dgageot commented:
I will change the name!
Le sam. 10 oct. 2015 à 21:07, Olivier Gambier <notifications@github.com> a
écrit :

> Looks good.
> @nathanleclaire <https://github.com/nathanleclaire> can you take a quick
> look?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/machine/pull/1960#issuecomment-147117619>.
>

dgageot commented:
@dmp42 test name fixed. If you or @nathanleclaire could take a look, that would be awsome

dmp42 commented:
LGTM

ping @nathanleclaire can you take a look?

nathanleclaire commented on commands/create_test.go:
```diff
@@ -20,3 +21,19 @@ func TestValidateSwarmDiscoveryAcceptsValidFormat(t *testing.T) {
 	assert.NoError(t, err)
 }
+
+func TestRejectCreateTrailingOptions(t *testing.T) {
```
This seems to be more of an integration test than a unit test (it tests the whole flow), so my preference would be to leave it out.

dgageot replied:
I strongly disagree @nathanleclaire. We should test most error cases without integration tests. This test might me a little complicated for now but it's not an integration test. It's just that the design has to be improved to make testing easier.

Until we do those improvements, that's one more test on a file that lacks tests

nathanleclaire commented:
Why change the `cmdCreate` method so much?  I have made some pretty serious changes in the whole `create` flow for the plugins so I'd prefer the changes to be minimal for now.

dgageot commented:
I changed `create` to make it testable. Having a dozen of `log.Fatal` is plain non-sense test-wise. It makes it impossible to unit-test.

nathanleclaire commented on test/integration/cli/create-rm.bats:
```diff
@@ -52,10 +52,10 @@ load ${BASE_TEST_DIR}/helpers.bash
+  run machine create -d none --url none a foo
   [ "$status" -eq 1 ]
-  [[ ${lines[0]} == "Host already exists: \"a\"" ]]
+  [[ ${lines[0]} == "Invalid command line. All options should be either before or after the machine name. Found those trailing flags [foo]" ]]
```
This error message doesn't make much sense to me as `foo` is not a flag, it's an extra argument.

dgageot replied:
Good point. I'll change the error message

dgageot commented:
@nathanleclaire I removed the modifications on `cmdCreate` since it's a pain for your PR. I still believe using `log.Fatal()` like this makes the code impossible to test with unit-tests so I'll create a PR later to address this issue.

I also fixed the wording.

Do you see any issue now?

nathanleclaire commented:
>  I removed the modifications on cmdCreate since it's a pain for your PR. I still believe using log.Fatal() like this makes the code impossible to test with unit-tests so I'll create a PR later to address this issue.

Yeah, I agree with this, and actually with the new plugins work anything we can do to reduce the possibility of a non-expected exit from the program is a good thing.

dgageot commented:
@nathanleclaire cool. So I guess it's ready to merge right? cc @nathanleclaire @dmp42

dmp42 commented:
Thanks @nathanleclaire ! 
And thanks @dgageot 

LGTM
